### PR TITLE
chore: add test terraform module

### PR DESCRIPTION
**Pull request by luke-zhu**
## What does this PR do?

This changes adds a test terraform module allows us to write some basic tests that inspect the output of successfully-created terraform modules.

## Testing

I ran the tests.

### CODE REVIEW

File: outputs.tf [hunk 1]
```diff
@@ -12,3 +12,9 @@ output "bucket" {
   description = "S3 bucket subscribed to Observe Lambda"
   value       = local.s3_bucket
 }
+
+# experimental: do not use yet
```
Comment (bendrucker) [inline outputs.tf:16]: Include this in the description, if anything. How would an output be experimental though, and what is the criteria for removal of that status?

File: test/main.tf [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Basic terraform tests. If they break, you probably have introduced a breaking change.
+# These tests cannot catch errors, but can be used to verify that the more complex variables worked as expected.
+
+provider "aws" {
+  region  = var.region
+  profile = var.profile
+}
+
+# Defaults, except no multi-region cloudtrail to avoid hitting 5 trail limit
+module "test_module_1" {
```
Comment (bendrucker) [inline test/main.tf:10]: Give these meaningful names, e.g., `module.test` and `module.test_all`, assuming the goals are a minimal test and a maximal test.

File: test/variables.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+variable "region" {
+  type        = string
+  description = "AWS Region"
+}
+
+variable "profile" {
+  type        = string
+  description = "AWS Profile"
+  default     = ""
```
Comment (bendrucker) [inline test/variables.tf:14]:
```suggestion
  default     = null
```

File: test/variables.tf [hunk 4]
```diff
@@ -0,0 +1,15 @@
+variable "name" {
+  type        = string
+  description = "Name"
+}
+
+variable "region" {
```
Comment (bendrucker) [inline test/variables.tf:6]: Region and profile can be set from env vars, consider whether you need these variables and explicit provider configuration.

File: Makefile [hunk 5]
```diff
@@ -6,4 +6,14 @@ changelog:
 release:
 	semtag final -s minor
 
+.PHONY: test
 test:
+	pre-commit run
+
+.PHONY: test-slow
```
Comment (bendrucker) [inline Makefile:13]: Remove "slow." Naming should identify the behavior, not describe it quantitatively. `test-single-region` and `test-all-regions` would be an appropriate alternative.

File: test/test.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        "eu-north-1",
+        "sa-east-1",
+    ]
+else:
+    regions = ["us-west-2"]
+
+for region in regions:
+    command = f"""
+terraform init -upgrade
+terraform apply -auto-approve -var='profile=thunderdome' -var="name=observe-${{USER}}" -var="region={region}"
```
Comment (bendrucker) [inline test/test.py:32]: Do not hardcode internal information (environments, accounts) in public modules, even in the tests

File: test/main.tf [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  name             = "${var.name}-1"
+  observe_customer = "123"
+  observe_token    = "ds123456789:dogdogdogdog"
+
+  cloudtrail_is_multi_region_trail = false
+}
+
+locals {
+  result1 = {
+    "cloudtrail" : length(module.test_module_1.cloudtrail) > 0,
```
Comment (bendrucker) [inline test/main.tf:21]: What is this trying to do? Checking for the `length` of a resource is unexpected and its behavior shouldn't be relied on. `terraform apply` would error if the trail resource could not be created, so when does this condition ever not pass?

File: test/test.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import subprocess
```
Comment (bendrucker) [inline test/test.py:1]: This functionality does not seem to justify the use of Python. A bash script would be able to handle this fine.

**Review by bendrucker — CHANGES_REQUESTED** [review]

File: test/main.tf (hunk 2 above)
Comment (luke-zhu) [reply test/main.tf:10]: done

File: outputs.tf (hunk 1 above)
Comment (luke-zhu) [reply outputs.tf:16]: i added it in the description. basically it's only used in a test. i'm not sure what our plans are for exposing almost all resources as outputs

File: test/test.py (hunk 6 above)
Comment (luke-zhu) [reply test/test.py:32]: ack

File: test/main.tf (hunk 7 above)
Comment (luke-zhu) [reply test/main.tf:21]: the cloudtrail resource involves `count` so there is either 1 or 0 elements

File: test/test.py (hunk 8 above)
Comment (luke-zhu) [reply test/test.py:1]: i just used what was easier for me

File: test/variables.tf (hunk 3 above)
Comment (luke-zhu) [reply test/variables.tf]: removed

File: test/variables.tf (hunk 4 above)
Comment (luke-zhu) [reply test/variables.tf]: this worked, removed

File: Makefile (hunk 5 above)
Comment (luke-zhu) [reply Makefile:13]: i think names have a lot of purposes, not just for identifying behavior.

i am not sure i want more concrete names like test-single-region if the implementation of the tests will likely change over time. The only structure I've put in place is for deciding where to place a test case based on how slow it is.

Down the road, the tests may be stable enough that we can start determining which tests people are expected to
run before submitting a change and monitor to ensure that those tests are fast enough.

File: test/test.py (hunk 8 above)
Comment (bendrucker) [reply test/test.py:1]: Ok, personal preference is not a good justification for tool choices. Optimize for readability and ease of use for other contributors, not yourself.

Shell is the default choice because it's portable and very easy to execute CLIs in a series of commands. If there were something especially difficult to handle in shell (concurrency, parsing, math, etc.), that's a good reason to introduce another tool.

File: test/main.tf (hunk 7 above)
Comment (bendrucker) [reply test/main.tf]: Ah right that makes a lot more sense. `length` with `for_each` or `count` is good. Question still stands re: `terraform apply` though. It seems like the goal is to test that the module was applied, not the conditional logic of trail creation.

File: Makefile (hunk 5 above)
Comment (bendrucker) [reply Makefile:13]: I am not persuaded. Names are an identifying label. Right now you have two test modes, single region and all region. If the tests change their intended behavior, the names should change too.

"Slow" is not an identifier for the declared tests, it is a descriptor of an emergent behavioral property, i.e. a side effect of how the tests are defined. 

In a huge test suite, you might have something like `--skip-slow` or `--skip-integration` etc, because the number of tests is so large that a caller could not realistically be expected to identify and skip them by name. In a suite of 2 tests, this is a premature abstraction.

File: test/test.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import os
+
+AWS_PROFILE = os.environ["AWS_PROFILE"]
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--slower", action="store_true", default=False)
+args = parser.parse_args()
+if args.slower:
+    regions = [
+        "us-east-1",
```
Comment (bendrucker) [inline test/test.py:12]: We can probably stand to drop most of these regions. Testing in `us-east-1`, `us-west-2`, and perhaps a few select global regions should be fine. 

Alternatively, `aws ec2 describe-regions` should be used if the intent is to test all regions, but that seems unnecessary. The differences between regions shouldn't be meaningful in most cases.

File: test/main.tf (hunk 7 above)
Comment (luke-zhu) [reply test/main.tf]: this test checks that the module input variables work as intended. We would expect a CloudTrail to be created by the module.

File: test/test.py (hunk 8 above)
Comment (luke-zhu) [reply test/test.py:1]: I don't think our team has decided on a default choice.

We use Python in a few GCP, Azure, and Linux host monitoring areas. It also seems like more members of our team are familiar with Python than Shell.

Shell is portable, but portability isn't the primary feature in this case. I think dev velocity is more important for now, so I leaned towards a tool that is easy for me to use but also doesn't raise to barrier to entry for others.

File: test/test.py (hunk 9 above)
Comment (luke-zhu) [reply test/test.py:12]: So far, I am noticing failures in different regions because some services aren't available in all regions. For example, RDS isn't in all regions.

I think it's valuable to be able to tell customers that our product works in all regions because we tested in all regions. We can't do that yet.

If we know it doesn't work in all regions, then we'd need to do back-and-forth communication, which is costly if it involves manual testing.

I am hoping this can become a CI test that takes ~30 minutes. However, it may be too expensive given how GitHub actions works.

File: Makefile (hunk 5 above)
Comment (luke-zhu) [reply Makefile:13]: sounds good, i can give concrete names here

File: test/test.py (hunk 9 above)
Comment (bendrucker) [reply test/test.py:12]: Ok, observed inter-region differences is a good reason to test across regions. Do make sure that's programmatic though, and not hardcoded. You can hardcode regions that should be excluded if there's known incompatibilities.

Comment (bendrucker) [reply test/test.py:12]: As far as Actions costs, the cost of a broken module is way higher than actions minutes, so don't worry about that. What I would worry about is whether someone has to wait 30 minutes to merge a trivial change. Often folks will use scheduled (nightly) builds for this purpose, so that regional edge cases are caught eventually, but development isn't slowed down everywhere to protect against a rare case. You can also offer a `workflow_dispatch` trigger and run the full suite for any potentially risky change that you identify as needing a full test run.

File: Makefile (hunk 5 above)
Comment (bendrucker) [inline Makefile:13]: This an the other PHONY below need updated target names

File: Makefile [hunk 10]
```diff
@@ -6,4 +6,14 @@ changelog:
...
+test-single-region: test
+	cd ./test/ &&	python3 test.py > test-single-region.txt 2>&1
```
Comment (bendrucker) [inline Makefile:15]: Writing test output to disk is unexpected, if there's a good reason to do this make sure it's included in `.gitignore`

File: s3.tf [hunk 11]
```diff
@@ -176,7 +176,17 @@ data "aws_iam_policy_document" "bucket" {
 
 }
 
+resource "time_sleep" "wait_before_subscribing" {
```
Comment (bendrucker) [inline s3.tf:179]: When did this sneak in? Since this is a user-facing change, land it as a separate PR.

File: test/test.py (hunk 8 above)
Comment (bendrucker) [reply test/test.py:1]: I'm flagging this not as a Python vs. Shell thing. At some point shell gets unwieldy and a proper language with better type support. I'd generally prefer to see Go, more because of well-established tools like https://pkg.go.dev/github.com/gruntwork-io/terratest than a fundamental language difference, but that's not the issue here.

The meat of this program, lines 34-43, is a shell program. The Python bits around it handle flag parsing and a loop. Both things that shell does well. If you were parsing the output of `output -json` into a dictionary, doing complex assertions, etc., I wouldn't ask you to shoehorn that into a shell script. But right now you're doing all the hard parts in a shell, and Python just ends up being a distraction for negligible benefit, if not net cost.

Consider, for example, how signals are handled. If you `SIGINT` a test run, does it terminate early and gracefully? In a shell, you can just change `terraform` to `exec terraform` and Terraform becomes the foreground process and receives signals directly. In an interpreted program which spawns a shell, you have multiple layers through which signals need to be propagated. Possible for sure, but shell has its strengths, and spawning CLIs in series and printing their output is shell's ultimate strength.

**Comment by luke-zhu** [comment]
resuming on this

File: Makefile (hunk 5 above)
Comment (luke-zhu) [reply Makefile]: thanks for catching, updated

File: Makefile (hunk 10 above)
Comment (luke-zhu) [reply Makefile]: removed writing to disk

File: s3.tf (hunk 11 above)
Comment (luke-zhu) [reply s3.tf]: commented this out, the tests often fail without this because of a race condition

Comment (luke-zhu) [reply s3.tf]: reducing a race condition with timeouts isn't optimal, but the cost of trying to remove the race condition doesn't seem worth it

**Comment by luke-zhu** [comment]
the tests seem to still hit race conditions, but terraform apply and destroy mostly work

File: test/test.py (hunk 8 above)
Comment (luke-zhu) [reply test/test.py]: i changed it to shell, seems to have been easier than i thought, thanks.

**Comment by luke-zhu** [comment]
not sure why terraform validate is failing:

│ Error: Unsupported block type
│ 
│   on ../eventbridge.tf line 30:
│   30: moved {
│ 
│ Blocks of type "moved" are not expected here.
╵

╷
│ Error: Unsupported block type
│ 
│   on ../eventbridge.tf line 30:
│   30: moved {
│ 
│ Blocks of type "moved" are not expected here.
╵

╷
│ Error: Unsupported block type
│ 
│   on ../firehose.tf line 23:
│   23: moved {
│ 
│ Blocks of type "moved" are not expected here.

File: test/test.sh [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+test_module_name=observe-${USER}
+INFO "TF Module Name: ${test_module_name}"
+
+
+AWS_REGION=$region terraform init -no-color
+
+for region in ${regions[@]}; do
+    DEBUG "Testing region $region"
+    AWS_REGION=$region terraform apply -auto-approve -no-color -var="name=${test_module_name}"
+    AWS_REGION=$region terraform output
```
Comment (bendrucker) [inline test/test.sh:69]: `export AWS_REGION=$region` would spare you the need to pass these to each `terraform` call.

File: test/test.sh [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    RESULT=$(terraform output -json result)
+    if [ "$RESULT" != "true" ]; then
+        echo "Test failed"
+        exit 1
+    fi
+    # Comment this to not destroy
```
Comment (bendrucker) [inline test/test.sh:75]: Provide config, e.g. a flag or `SKIP_DESTROY`, avoid telling users to mutate the script. Maybe even document this in the output, e.g.,

```
echo "Running terraform destroy, set SKIP_DESTROY=1 to skip this step"
```

File: versions.tf [hunk 14]
```diff
@@ -11,5 +11,13 @@ terraform {
       source  = "hashicorp/random"
       version = ">= 3.0.0"
     }
+
+    # TODO(luke): uncomment this in a follow up commit to make terraform-aws-collection
+    # more likley to avoid the terraform apply failure
+    # (due to s3 bucket notification eventual consistency)
+    time = {
```
Comment (bendrucker) [inline versions.tf:18]: Seems like this is out of sync with the rest? Meaning if this provider isn't used, the requirement should be commented out?

File: s3.tf [hunk 15]
```diff
@@ -184,7 +184,24 @@ data "aws_iam_policy_document" "bucket" {
 
 }
 
+# TODO(luke): uncomment this in a follow up commit to make terraform-aws-collection
+# more likley to avoid the terraform apply failure
+# (due to s3 bucket notification eventual consistency)
+#
+# resource "time_sleep" "wait_before_subscribing" {
+#   depends_on = [
```
Comment (bendrucker) [inline s3.tf:192]: `depends_on` must refer to a specific resource or module address and cannot accept arbitrary expressions

**Review by bendrucker — COMMENTED** [review]
Nice! Few remaining comments about this race condition. The script is a nice improvement, appreciate all the colorized and debug output.

**Comment by bendrucker** [comment]
Re `validate`, you may need to rebase this, things like `nullable` have been added recently as breaking changes and I think the min TF version has been incremented.

File: s3.tf (hunk 15 above)
Comment (luke-zhu) [reply s3.tf:192]: done

File: versions.tf (hunk 14 above)
Comment (luke-zhu) [reply versions.tf:18]: done

**Comment by luke-zhu** [comment]
maybe the min tf version doesn't support `move` and `nullable`

**Comment by luke-zhu** [comment]
seems like `moved` was introduced in 1.1, while the min in the Action is 1.0
